db/range_del_aggregator.cc

```diff
@@ -138,159 +139,80 @@ class CollapsedRangeDelMap : public RangeDelMap {
            ucmp_->Compare(iter_->first, parsed.user_key) <= 0);
     assert(std::next(iter_) == rep_.end() ||
            ucmp_->Compare(parsed.user_key, std::next(iter_)->first) < 0);
-    return parsed.sequence < iter_->second.seq_;
+    return parsed.sequence < iter_->second;
   }
 
   bool IsRangeOverlapped(const Slice&, const Slice&) {
     // so far only implemented for non-collapsed mode since file ingestion (only
     // client) doesn't use collapsing
     assert(false);
+    return false;
   }
 
-  void AddTombstone(RangeTombstone tombstone) {
-    // In collapsed mode, we only fill the seq_ field in the RangeDelMap's
-    // values. The end_key is unneeded because we assume the tombstone extends
-    // until the next tombstone starts. For gaps between real tombstones and
-    // for the last real tombstone, we denote end keys by inserting fake
-    // tombstones with sequence number zero.
-    std::vector<RangeTombstone> new_range_dels{
-        tombstone, RangeTombstone(tombstone.end_key_, Slice(), 0)};
-    auto new_range_dels_iter = new_range_dels.begin();
-    // Position at the first overlapping existing tombstone; if none exists,
-    // insert until we find an existing one overlapping a new point
-    const Slice* rep_begin = nullptr;
-    if (!rep_.empty()) {
-      rep_begin = &rep_.begin()->first;
-    }
-    auto last_range_dels_iter = new_range_dels_iter;
-    while (new_range_dels_iter != new_range_dels.end() &&
-           (rep_begin == nullptr ||
-            ucmp_->Compare(new_range_dels_iter->start_key_, *rep_begin) < 0)) {
-      rep_.emplace(new_range_dels_iter->start_key_,
-                   RangeTombstone(Slice(), Slice(), new_range_dels_iter->seq_));
-      last_range_dels_iter = new_range_dels_iter;
-      ++new_range_dels_iter;
-    }
-    if (new_range_dels_iter == new_range_dels.end()) {
-      return;
+  void AddTombstone(RangeTombstone t) {
+    auto it = rep_.upper_bound(t.start_key_);
+    auto prev_seq = [&]() {
```

Comment:
I can't remember the rules for whether this heap allocates or not. It might be better to pass `it` as a parameter to this functor as I'm pretty sure that won't allocate.

Reply:
Lambdas don't heap allocate unless you `new` them according to http://www.elbeno.com/blog/?p=1068 and https://stackoverflow.com/questions/12202656/c11-lambda-implementation-and-memory-model.) Clang seems to have no trouble inlining this either based on my read of the generated assembly.

Reply:
Thanks for the pointer.